crypto/box/src/hkdf_box.rs

```diff
@@ -0,0 +1,169 @@
+
+        let curve_point_bytes =
+            GenericArray::<u8, <KexAlgo::Public as PublicKey>::Size>::clone_from_slice(
+                our_public.get_bytes().as_ref(),
+            );
+
+        // KDF
+        let (aes_key, aes_nonce) = Self::kdf_step(&shared_secret);
+
+        // AES
+        let aead = AeadAlgo::new(aes_key);
+        let mac = aead.encrypt_in_place_detached(&aes_nonce, &[], buffer)?;
+
+        // Tag is curve_point_bytes || aes_mac_bytes
+        Ok(curve_point_bytes.concat(mac))
+    }
+
+    fn decrypt_in_place_detached(
+        &self,
+        key: &KexAlgo::Private,
+        tag: &GenericArray<u8, Self::FooterSize>,
+        buffer: &mut [u8],
+    ) -> Result<(), Error> {
+        // ECDH
+        use keys::KexReusablePrivate;
+        let public_key =
+            KexAlgo::Public::try_from(&tag[..<KexAlgo::Public as PublicKey>::Size::USIZE])
+                .map_err(Error::Key)?;
+        let shared_secret = key.key_exchange(&public_key);
+
+        // KDF
+        let (aes_key, aes_nonce) = Self::kdf_step(&shared_secret);
+
+        // AES
+        let mac_ref = <&GenericArray<u8, AeadAlgo::TagSize>>::from(
+            &tag[<KexAlgo::Public as PublicKey>::Size::USIZE..],
+        );
+        let aead = AeadAlgo::new(aes_key);
+        aead.decrypt_in_place_detached(&aes_nonce, &[], buffer, mac_ref)
+            .map_err(|_| Error::MacFailed)?;
+
+        Ok(())
+    }
+}
+
+impl<KexAlgo, DigestAlgo, AeadAlgo> HkdfBox<KexAlgo, DigestAlgo, AeadAlgo>
+where
+    KexAlgo: Kex,
+    for<'privkey> <KexAlgo as Kex>::Public: From<&'privkey <KexAlgo as Kex>::EphemeralPrivate>,
+    DigestAlgo: Digest + Input + FixedOutput + Default + Clone + BlockInput + Reset,
+    AeadAlgo: Aead + NewAead,
+    AeadAlgo::NonceSize:
+        IsLessOrEqual<<KexAlgo::Secret as KexSecret>::EntropyLowerBound, Output = B1>,
+    AeadAlgo::KeySize:
+        IsLessOrEqual<<KexAlgo::Secret as KexSecret>::EntropyLowerBound, Output = B1>,
+    <<KexAlgo as Kex>::Public as PublicKey>::Size: Add<<AeadAlgo as aead::Aead>::TagSize>,
+    Sum<<KexAlgo::Public as PublicKey>::Size, AeadAlgo::TagSize>: ArrayLength<u8>,
+{
+    /// KDF part, factored out to avoid duplication
+    /// This part must produce the key and IV/nonce for Aead, from the IKM, using Hkdf.
+    fn kdf_step(
+        dh_secret: &KexAlgo::Secret,
+    ) -> (
+        GenericArray<u8, AeadAlgo::KeySize>,
+        GenericArray<u8, AeadAlgo::NonceSize>,
+    ) {
+        let kdf = Hkdf::<DigestAlgo>::new(Some(b"dei-salty-box"), dh_secret.as_ref());
+        let mut key: GenericArray<u8, AeadAlgo::KeySize> = Default::default();
+        let mut nonce: GenericArray<u8, AeadAlgo::NonceSize> = Default::default();
+        kdf.expand(b"aead-key", key.as_mut_slice())
+            .expect("Aead::KeySize is too large compared to Digest output size");
+        kdf.expand(b"aead-nonce", nonce.as_mut_slice())
+            .expect("Aead::NonceSize is too large compared to Digest output size");
```

Comment:
Not sure if you intended this, but nonce and key will not be successive expansions on an underlying state, it will run the same expansion twice, with different info parameters. If you want to derive multiple keys in the "usual" KDF way, you need to call expand with a buffer that is as long as both your key/nonce combined, and splice them.

Now that I'm thinking about it, we could probably use a [one-way Noise handshake](https://noiseprotocol.org/noise.html#one-way-handshake-patterns) to accomplish all this...

Reply:
thanks, i didn't think of this, i was mostly just mimicking how we do the account key derivation. there was something I saw online that suggested using hkdf this way... i'm going to try to find a link

Reply:
> we could probably use a one-way Noise handshake to accomplish all this...

that would be cool! i'm sure if you have embedded devices that already want to use noise for secure channels, they would appreciate a smaller footprint from not having hkdf?

i'm not sure if we need hkdf anywhere else in any of the enclaves?

Reply:
also, bonus points for making a type called "noise box"

Reply:
@jcape does the `mcnoise` crate have a dependency on `Vec` / `alloc`? Cause right now, `cryptobox` and `hkdf_box` do not. That would be a reason to keep hkdf box around

Reply:
yah it looks to me like it does

if you want for like, a tiny off-line device to be able to construct transactions, including fog hints, then it might not want to have noise as a dependency?

Reply:
i think its better to have a simple proof of security than a complicated one... the one in the readme for hkdf box is like a few lines -- the statement being proved in the paper you linked is like vastly more complicated and proves something far more powerful. also it makes it clear that different properties are reached at different stages. i can't figure out what it guarantees for my scenario without reading the whole paper

Reply:
My thinking on this is simple: a published proof is easier to trust than an unpublished one, an existing scheme is easier to trust than a novel one, and audited code is easier to trust than un-audited code. If/when vec becomes a problem, we can replace it, `aead` has in-place APIs to handle things now.